crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -169,6 +188,17 @@ impl CheckpointExecutor {
             assert_eq!(seq, 0);
         }
         debug!("Bumping highest_executed_checkpoint watermark to {:?}", seq,);
+        let (callback_sender, mut callback_receiver) = mpsc::channel(SUBSCRIPTION_CAPACITY);
+        match self.subscriptions.send((execution_state, callback_sender)) {
+            Ok(num_subscriptions) => {
+                let mut received_callbacks = 0;
+                while callback_receiver.recv().await.is_some() {
```

Comment:
So essentially here we wait for pruner to finish before we continue with checkpoint execution - what is the reason for this?

Reply:
great question! yeah, not only pruner but any other potential checkpoint subscriber(i.e. state accumulator, etc.)
The main reason is the much easier standard crash recovery. Subscribers won't need to keep their own state, just keep tasks idempotent.
Of course it has a tradeoff of added latency for checkpoint execution. But FWIU it's tolerable for now and preferable for the sake of simplicity
maybe @mystenmark can elaborate more

Reply:
I see. I don't think I agree with trade off(it seem that developing generic interface for subscriber to just track last version is not super complex and avoids having non-critical tasks such as pruner on the critical path of checkpoint execution), but it's your call. For the very least I would measure how much time is spent waiting for callbacks (you can use `monitored_scope` for that).

Reply:
Agree that we should at least measure the time spent in the callbacks to be able to track the ones stalling checkpoint execution.

Reply:
added `monitored_scope`
as for implementation, I can definitely add dedicated task that tracks progress for subscribers and resends missing checkpoints. It also will have to store progress in db to handle process crash recovery. One caveat though is that not all subscribers are like pruner. For pruner it's fine to be lagging behind the most recent executed checkpoint. Not sure how about things like StateAccumulator.
But I totally get added latency concern.
We can start with the simple approach, monitor latency and go from there

Reply:
For `StateAccumulator`, we most likely will also need to do things non-blocking. It turns out accumulating checkpoints has non-negligible latency.

If we want a single point for crash recovery but want to ensure that things are non-blocking, I think checkpoint executor is the place to enforce that. And the `FuturesOrdered` queue already gives these semantics. To that end, couldn't you just move the blocking send to the pruner into the execution task? This would mean that we don't bump the `highest_executed_checkpoint` watermark until after pruning (and could adopt the same strategy for state accumulator). Then in crash recovery, we would just reschedule the checkpoint, which would cause re-prune, and re-accumulation (so just need to make sure these operations are idempotent). 

Or would pruning out of order cause issues?

Reply:
@mlogan: I can make it non-generic for now, specially cause it looks like state accumulator's case is a bit different from pruner. But FWIU archival solution will need checkpoint subscription too

Pruning is idempotent, but ordering definitely matters for pruner, so I don't think execution task is the place to handle it

Reply:
changed to non-generic implementation for now. Will return broadcast model if we get more subscribers